kirkwood: use nvmem for mac

neheb opened the pull request:
Userspace handling is deprecated.

ping @mans0n

neheb committed:
kirkwood: use nvmem for mac

Userspace handling is deprecated.

Signed-off-by: Rosen Penev <rosenp@gmail.com>

mans0n commented:
Thanks for taking care of this. Actually I had also tried it before, but it didn't work.

In the early boot sequence, `linux/arch/arm/mach-mvebu/kirkwood.c:kirkwood_dt_eth_fixup()` calls `of_get_mac_address()`, but at that time the mtd is not initialized yet, so it fails to load the nvmem-cells. Instead, it loads another MAC address from Marvell (which is different from the one provided by the device vendor) and set it to local-mac-address.
After the Ethernet driver is loaded, `of_get_mac_address()` is called again, and it prefers local-mac-address over the nvmem-cells.

... and this is what I memoed in 2022 so the internal things might be different now, but I've just tried it again and I confirm it still don't work. :/

neheb commented:
Unfortunate. There's

https://git.kernel.org/pub/scm/linux/kernel/git/stable/linux.git/commit/drivers/net/ethernet/marvell/mv643xx_eth.c?h=v6.6.51&id=11f8e7c122ce013fa745029fa8c94c6db69c2e54 has the necessary change. But if it's not working there's no point.

neheb commented:
Looking closer, kirkwood_dt_eth_fixup seems outdated. Everything looks like it has dts.

neheb committed:
kirkwood: allow overriding mac address with nvmem

Because this platform predates dts, there is a hack to use
local-mac-address and prefer it over nvmem. Get rid of it.

Signed-off-by: Rosen Penev <rosenp@gmail.com>

neheb commented:
@mans0n how about now?

neheb commented:
Rebased after kernel 6.1 removal.

hauke commented on target/linux/kirkwood/patches-6.6/300-nvmem.patch:
```diff
@@ -0,0 +1,107 @@
+-
+-		/* skip disabled nodes or nodes with valid MAC address*/
+-		if (!of_device_is_available(pnp) ||
+-		    !of_get_mac_address(np, tmpmac))
```
I think removing this will not be accapted upstream.

You could try to check if `of_get_mac_address()` returns `-EPROBE_DEFER` and then go to `eth_fixup_skip` here. In such a case it should get a correct mac address later.

neheb replied:
This is in early init. No way they would accept that.

neheb replied:
@hauke anyway, made the change. I'll try upstreaming but I'm skeptical.

neheb replied:
Oh wait this is completely wrong. You're confusing of_get_mac_address with of_get_etherdev_address.

hauke commented:
@mans0n  or someone else with this device: Could you please test this and report back if it is working.

neheb commented:
Rebased and refreshed patch to pass CI.

mans0n commented:
That did fetch the correct LAN and label mac address, with this modification (`&eth1 -> &eth1port`):

```diff
--- a/target/linux/kirkwood/files-6.6/arch/arm/boot/dts/marvell/kirkwood-nas1.dts
+++ b/target/linux/kirkwood/files-6.6/arch/arm/boot/dts/marvell/kirkwood-nas1.dts
@@ -16,6 +16,7 @@
 		led-failsafe = &led_ready;
 		led-running = &led_ready;
 		led-upgrade = &led_ready;
+		label-mac-device = &eth1port;
 	};
 
 	chosen {
@@ -218,6 +229,9 @@
 &eth1port {
 	phy-handle = <&ethphyb>;
 	phy-connection-type = "rgmii-id";
+
+	nvmem-cells = <&macaddr_uboot_3ffa8>;
+	nvmem-cell-names = "mac-address";
 };
 
 &sata {
```

mans0n commented:
And I also doubt that the kirkwood_dt_eth_fixup patch will be accepted upstream, that's why I chose the userspace way at the first place... But we can try.

neheb commented:
Yeah this patch effectively makes the eth fixup function a noop if CONFIG_NVMEM is enabled.

I wonder if it breaks other kirkwood devices.
